pkg/provision/gce.go

```diff
@@ -0,0 +1,210 @@
+	}
+
+	status := ""
+
+	if op.Status == "RUNNING" {
+		status = ActiveStatus
+	}
+	return &ProvisionedHost{
+		ID:     constructCustomGCEID(host.Name, host.Additional["zone"], host.Additional["projectid"]),
+		Status: status,
+	}, nil
+
+}
+
+// CheckInletsFirewallRuleExists checks if the inlets firewall rule exists or not
+func (gce *GCEProvisioner) CheckInletsFirewallRuleExists(id string, inletsPort string) (bool, error) {
+	_, _, projectID, err := getGCEFieldsFromID(id)
+	if err != nil {
+		return false, fmt.Errorf("Could not get custom GCE fields: %v", err)
+	}
+	firewallRuleName := "inlets"
+	op, err := gce.gceProvisioner.Firewalls.Get(projectID, firewallRuleName).Do()
+	if err != nil {
+		return false, fmt.Errorf("Could not get inlets firewall rule: %v", err)
+	}
+	if op.Name == firewallRuleName {
+		for _, firewallRule := range op.Allowed {
+			for _, port := range firewallRule.Ports {
+				if port == inletsPort {
+					return true, nil
+				}
+			}
+		}
+	}
+	return false, nil
+}
+
+// CreateInletsFirewallRule creates a firewall rule opening up the control port for inlets
+func (gce *GCEProvisioner) CreateInletsFirewallRule(id string, inletsPort string) error {
+	_, _, projectID, err := getGCEFieldsFromID(id)
+	if err != nil {
+		return fmt.Errorf("Could not get custom GCE fields: %v", err)
+	}
+	firewallRuleName := "inlets"
+	firewallRule := &compute.Firewall{
+		Name:    firewallRuleName,
+		Network: fmt.Sprintf("projects/%s/global/networks/default", projectID),
+		Allowed: []*compute.FirewallAllowed{
+			{
+				IPProtocol: "tcp",
+				Ports:      []string{inletsPort},
+			},
+		},
+		SourceRanges: []string{"0.0.0.0/0"},
+		Direction:    "INGRESS",
+		TargetTags:   []string{"inlets"},
+	}
+
+	_, err = gce.gceProvisioner.Firewalls.Insert(projectID, firewallRule).Do()
+	if err != nil {
+		return fmt.Errorf("could not create firewall rule: %v", err)
+	}
+	return nil
+}
+
+// Delete deletes the GCE exit node
+func (gce *GCEProvisioner) Delete(id string) error {
+	instanceName, zone, projectID, err := getGCEFieldsFromID(id)
+	if err != nil {
+		return fmt.Errorf("Could not get custom GCE fields: %v", err)
+	}
+	_, err = gce.gceProvisioner.Instances.Delete(projectID, zone, instanceName).Do()
+	if err != nil {
+		return fmt.Errorf("Could not delete the GCE instance: %v", err)
+	}
+	return err
+}
+
+// Status checks the status of the provisioning GCE exit node
+func (gce *GCEProvisioner) Status(id string) (*ProvisionedHost, error) {
+	instanceName, zone, projectID, err := getGCEFieldsFromID(id)
+	if err != nil {
+		return nil, fmt.Errorf("Could not get custom GCE fields: %v", err)
+	}
+
+	op, err := gce.gceProvisioner.Instances.Get(projectID, zone, instanceName).Do()
+	if err != nil {
+		return nil, fmt.Errorf("Could not get instance: %v", err)
+	}
+
+	status := ""
+
+	if op.Status == "RUNNING" {
+		status = ActiveStatus
+	}
+
+	return &ProvisionedHost{
+		IP:     op.NetworkInterfaces[0].AccessConfigs[0].NatIP,
```

Comment:
Have you thought about applying range-checking here? What if there are 0 rather than > 0 entries in both slices/arrays?

Reply:
I guess this won't be an issue because in the instance template struct I am requesting only a single NatIP, so either the provisioning call will succeed or it would fail completely.